src/dodal/devices/oav/oav_to_redis_forwarder.py

```diff
@@ -66,68 +75,70 @@ def __init__(
-        # callbacks to correlate the data
-        self.uuid, self.uuid_setter = soft_signal_r_and_setter(str)
+        with self.add_children_as_readables():
+            # The uuid that images are being saved under, this should be monitored for
+            # callbacks to correlate the data
+            self.uuid, self.uuid_setter = soft_signal_r_and_setter(str)
 
         super().__init__(name=name)
 
-    async def _get_frame_and_put_to_redis(self, response: ClientResponse):
-        """Converts the data that comes in as a jpeg byte stream into a numpy array of
-        RGB values, pickles this array then writes it to redis.
-        """
+    async def _get_frame_and_put_to_redis(
+        self, redis_uuid: str, response: ClientResponse
+    ):
         jpeg_bytes = await get_next_jpeg(response)
-        self.uuid_setter(image_uuid := str(uuid.uuid4()))
-        img = Image.open(io.BytesIO(jpeg_bytes))
-        image_data = pickle.dumps(np.asarray(img))
-        sample_id = str(await self.sample_id.get_value())
-        await self.redis_client.hset(sample_id, image_uuid, image_data)  # type: ignore
-        await self.redis_client.expire(sample_id, timedelta(days=self.DATA_EXPIRY_DAYS))
-        LOGGER.debug(f"Sent frame to redis key {sample_id} with uuid {image_uuid}")
+        self.uuid_setter(redis_uuid)
+        sample_id = await self.sample_id.get_value()
+        redis_key = f"murko:{sample_id}:raw"
+        await self.redis_client.hset(redis_key, redis_uuid, jpeg_bytes)  # type: ignore
+        await self.redis_client.expire(redis_key, timedelta(days=self.DATA_EXPIRY_DAYS))
 
     async def _open_connection_and_do_function(
-        self, function_to_do: Callable[[ClientResponse, str | None], Awaitable]
+        self, function_to_do: Callable[[ClientResponse, OAVSource], Awaitable]
     ):
-        source = await self.selected_source.get_value()
-        stream_url = await self._sources[source.value].get_value()
+        source_name = await self.selected_source.get_value()
+        LOGGER.info(
+            f"Forwarding data from sample {await self.sample_id.get_value()} and OAV {source_name}"
+        )
+        source = self._sources[source_name.value]
+        stream_url = await source.url.get_value()
         async with ClientSession() as session:
             async with session.get(stream_url) as response:
-                await function_to_do(response, stream_url)
+                await function_to_do(response, source)
 
-    async def _stream_to_redis(self, response, _):
-        while not self._stop_flag:
-            await self._get_frame_and_put_to_redis(response)
-            await asyncio.sleep(0.01)
+    async def _stream_to_redis(self, response: ClientResponse, source: OAVSource):
+        done_status = AsyncStatus(
+            asyncio.wait_for(self._stop_flag.wait(), timeout=self.TIMEOUT)
+        )
+        async for image_uuid in observe_value(self.counter, done_status=done_status):
```

Comment:
this counter signal, is it really a counter? It looks like it's actually uuids? Or are they not really uuids and are just sequence numbers?

Reply:
Q: How frequently does this signal update, is it guaranteed that we will see every frame id that we need to collect?

Reply:
This is the frame counter on the device. We can use it as a UUID as it increments with every frame and we're only using the one camera so will be unique. I suspect there is a roll-over at something around the max int value but given we don't keep data more than a week I think it's unlikely to affect us. 

It updates at the rate of the camera (~15Hz). You're right that we may end up missing some, though experimentally we haven't yet, but even if we do it's not critical we catch every frame.

Legitimate questions though, thanks for raising them, I will add some comments in the code to this affect.

Reply:
Re: uniqueness of the UUID

Unsure if it is overkill but can envisage during some operational issue a few camera IOC restarts could be needed. It is possible the counter always starts at one from IOC restart. In a combination of IOC restarts plus same sampleID being loaded and data collected on the same frame number (very unlikely combinations) then a UUID collision could technically happen. A possible solution would be to add a prefix of some kind randomly generated.

But unsure if it is worth the effort.

Reply:
Good point, thanks. I will add a proper UUID section in too as a precaution.